File: 116. Populating Next Right Pointers in Each Node/116. Populating Next Right Pointers in Each Node.py
class Solution(object):
    def connect(self, root):
        queue = [root]

        while queue:
            for i in range(len(queue) - 1):
                queue[i].next = queue[i+1]
           
            for _ in range(len(queue)):
                current = queue.pop(0)
                if current and current.left:
                    queue.append(current.left)
                    queue.append(current.right)
        return root
            

File: 116. Populating Next Right Pointers in Each Node/116. Populating Next Right Pointers in Each Node.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}
    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}
    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/
class Solution {
public:
    Node* connect(Node* root) {
        if (!root) return nullptr;
        
        Node* levelStart = root;
        while (levelStart->left) {  // while we have a next level
            Node* current = levelStart;
            while (current) {
                current->left->next = current->right;
                if (current->next) {
                    current->right->next = current->next->left;
                }
                current = current->next;
            }
            levelStart = levelStart->left;
        }
        
        return root;
    }
};